content/blog/2019/11/2019-11-21-welcome-to-the-matrix.adoc

```diff
@@ -0,0 +1,389 @@
+---
+layout: post
+title: "Welcome to the Matrix"
+tags:
+- pipeline
+- plugins
+- declarative
+- matrix 
+author: lnewman
+---
+
+I often find myself needing to run the same actions on a bunch of different configurations.  
+Up to now, that meant I had to make multiple copies of the same stages in my pipelines. 
+When I needed to make changes, I had to make the same changes in multiple places throughout my pipeline.
+Maintaining even a small number of configuration was difficult for larger pipelines.
+
+Declarative Pipeline 1.5.0-beta1 (now available from the
+link:https://updates.jenkins.io/experimental/[Jenkins Experimental Update site]) adds a new `matrix` section that lets me specify a list stages once and then run that same list in parallel on multiple configurations. Let's take a look!
+
+== Single configuration pipeline
+
+I'll start with a simple pipeline with build and test stages.  I'm using `echo` steps as placeholders for my build and test actions.
+
+.Jenkinsfile
+[source, groovy]
+----
+pipeline {
+    agent none
+    stages {
+        stage('BuildAndTest') {
+            agent any
+            stages {
+                stage('Build') {
+                    steps {
+                        echo 'Do Build'
+                    }
+                }
+                stage('Test') {
+                    steps {
+                        echo 'Do Test'
+                    }
+                }
+            }
+        }
+    }
+}
+----
+
+== Pipeline for multiple platforms and browsers
+
+I'd like to run my build and tests on a combination of platforms and browsers. 
+The new `matrix` directive lets me specify a set of `axes`. 
+Each `axis` has a `name` and a list of one or more `values`.
+When the pipeline is run, Jenkins will take those and run my stages on all possible combinations of values from each axis.
+All cells in a matrix run in parallel (limited only by the number of available agents).  
+Stages within each cell are run sequentially.
+
+My matrix has two axes: `PLATFORM` and `BROWSER`.
+I have three values for `PLATFORM` and four values for `BROWSER` resulting in a my stages being run with twelve different combinations being run.
```

Comment:
Suggested reword:

I have three values for `PLATFORM` and four values for `BROWSER`, resulting in my stages being run in twelve different combinations.

Reply:
```suggestion
I have three values for `PLATFORM` and four values for `BROWSER` resulting in a my stages being run with twelve different combinations being run.
```